source/MagicInternetBox.cpp

```diff
@@ -121,6 +123,87 @@ void MagicInternetBox::sendData(NetworkDataType type, float angle, int id, int d
 	ws->sendBinary(data);
 }
 
+void MagicInternetBox::syncState(std::shared_ptr<ShipModel> state) {
+	std::vector<uint8_t> data;
+	data.push_back(StateSync);
+
+	const auto& doors = state->getDoors();
+	data.push_back((uint8_t)doors.size());
+	for (unsigned int i = 0; i < doors.size(); i++) {
+		if (doors[i]->getAngle() == -1) {
+			data.push_back(0);
+			data.push_back(0);
+			data.push_back(0);
+		} else {
+			data.push_back(1);
+
+			int d3 = (int)(FLOAT_PRECISION * doors[i]->getAngle());
+			data.push_back((uint8_t)(d3 % ONE_BYTE));
+			data.push_back((uint8_t)(d3 / ONE_BYTE));
+		}
+	}
+
+	const auto& breaches = state->getBreaches();
+	data.push_back((uint8_t)breaches.size());
+	for (unsigned int i = 0; i < breaches.size(); i++) {
+		data.push_back(breaches[i]->getHealth());
+		data.push_back(breaches[i]->getPlayer());
+
+		int d3 = (int)(FLOAT_PRECISION * breaches[i]->getAngle());
+		data.push_back((uint8_t)(d3 % ONE_BYTE));
+		data.push_back((uint8_t)(d3 / ONE_BYTE));
+	}
+	ws->sendBinary(data);
+}
+
+void MagicInternetBox::resolveState(std::shared_ptr<ShipModel> state,
+									const std::vector<uint8_t>& message) {
+	const auto& doors = state->getDoors();
+	if (doors.size() != message[1]) {
+		CULog("ERROR: DOOR ARRAY SIZE DISCREPANCY; local %d but server %d", doors.size(),
+			  message[1]);
+		return;
+	}
+	int doorIndex = 2;
+	for (unsigned int i = 0; i < doors.size(); i++) {
+		if (message[doorIndex]) {
+			float angle = (float)(message[doorIndex + 1] + ONE_BYTE * message[doorIndex + 2]) /
+						  FLOAT_PRECISION;
+			if (abs(doors[i]->getAngle() - angle) > FLOAT_EPSILON) {
+				CULog("Found open door that should be closed, id %d", i);
+				state->createDoor(angle, (int)i);
+			}
+		} else {
+			if (doors[i]->getAngle() != -1.0f) {
+				CULog("Found closed door that should be open, id %d", i);
+				state->flagDoor((int)i, 0, 1);
+				state->flagDoor((int)i, 1, 1);
+			}
```

Comment:
@ssorenson130 Can you review this part in particular? Is there a better way to forcefully open a door that was closed?

Reply:
yes, this won't even really work for more than 2 players. just call `clear()` on the door

Reply:
The `clear()` method doesn't set the angle to `-1`

Reply:
ah I misread this. what you added works fine